srtcore/group.cpp

```diff
@@ -2969,6 +2969,112 @@ void CUDTGroup::sendBackup_CheckIdleTime(gli_t w_d)
+        m_fout << u.id() << ",";
+        m_fout << weight << ",";
+        m_fout << u.peer_latency_us() << ",";
+        m_fout << u.RTT() << ",";
+        m_fout << u.RTTVar() << ",";
+        m_fout << stability_tmo_us << ",";
+        m_fout << count_microseconds(currtime - u.LastRspTime()) << ",";
+        m_fout << state << ",";
+        m_fout << (srt::sync::is_zero(u.FreshActivationStart()) ? -1 : (count_microseconds(currtime - u.FreshActivationStart()))) << ",";
+        m_fout << activation_period_us << "\n";
+        m_fout.flush();
+    }
+
+private:
+    void print_header()
+    {
+        //srt::sync::ScopedLock lck(m_mtx);
+        m_fout << "Timepoint,SocketID,weight,usLatency,usRTT,usRTTVar,usStabilityTimeout,usSinceLastResp,State,usSinceActivation,usActivationPeriod\n";
+    }
+
+    void create_file()
+    {
+        if (m_fout)
+            return;
+
+        std::string str_tnow = srt::sync::FormatTimeSys(srt::sync::steady_clock::now());
+        str_tnow.resize(str_tnow.size() - 6); // remove trailing ' [SYS]' part
+        while (str_tnow.find(':') != std::string::npos) {
+            str_tnow.replace(str_tnow.find(':'), 1, 1, '_');
+        }
+        const std::string fname = "stability_trace_" + str_tnow + ".csv";
+        m_fout.open(fname, std::ofstream::out);
+        if (!m_fout)
+            std::cerr << "IPE: Failed to open " << fname << "!!!\n";
+
+        print_header();
+    }
+
+private:
+    srt::sync::Mutex m_mtx;
+    std::ofstream m_fout;
+};
+
+StabilityTracer s_stab_trace;
+#endif
+
+/// TODO: Remove 'weight' parameter? Only needed for logging.
+/// @retval  1 - link is identified as stable
+/// @retval  0 - link state remains unchanged (too early to identify, still in activation phase)
+/// @retval -1 - link is identified as unstable
+static int sendBackup_CheckRunningLinkStable(const CUDT& u, const srt::sync::steady_clock::time_point& currtime, uint16_t weight)
+{
+    const uint32_t latency_us = u.peer_latency_us();
+    const int32_t min_stability_us     = 60000; // Minimum Link Stability Timeout: 60ms.
+    const int64_t initial_stabtout_us  = max<int64_t>(min_stability_us, latency_us);
+    const int64_t activation_period_us = initial_stabtout_us + 5 * CUDT::COMM_SYN_INTERVAL_US;
+
+    // RTT and RTTVar values are still being refined during activation period,
+    // therefore the dymanic timeout should not be used in activation phase.
+    const bool is_activation_phase = !is_zero(u.FreshActivationStart())
+        && (count_microseconds(currtime - u.FreshActivationStart()) <= activation_period_us);
+
+    const int peer_idle_tout_us = u.peer_idle_tout_ms() * 1000;
+
+    const int64_t stability_tout_us = is_activation_phase
+        ? initial_stabtout_us // activation phase
+        : min<int64_t>(max<int64_t>(min_stability_us, 2 * u.RTT() + 4 * u.RTTVar()), latency_us);
+    
+    const steady_clock::duration td_response = currtime - u.LastRspTime();
```

Comment:
What if `u.lastRspTime() == 0` ?

Reply:
Also, what if `u.LastRspTime() < u.FreshActivationStart()` ?

Reply:
> What if `u.lastRspTime() == 0`?

It can't be zero. It is initialized with the current time when a socket is created or initially configured.

> what if `u.LastRspTime() < u.FreshActivationStart()`?

That's a good observation! Something like this should be done here then:

```c++
const steady_clock::time_point last_rsp = max(u.freshActivationStart(), u.lastRspTime());
const steady_clock::duration td_response = currtime - last_rsp;
```